awx/ui_next/src/screens/Template/WorkflowJobTemplateVisualizer/Modals/NodeModal/NodeNextButton.jsx

```diff
@@ -0,0 +1,39 @@
+import React, { useEffect } from 'react';
+import { func, number, shape, string } from 'prop-types';
+import { Button } from '@patternfly/react-core';
+
+function NodeNextButton({
+  activeStep,
+  buttonText,
+  onClick,
+  onNext,
+  triggerNext,
+}) {
+  useEffect(() => {
+    if (!triggerNext) {
```

Comment:
I traced how click events are handled between this component and its parent, `NodeModal` and as far as I can tell it works like this:

- `NodeNexButton` is clicked, which calls `onClick(activeStep)`
- The implementation of `onClick` inside `NodeModal` doesn't  accept any arguments though, so `activeStep` is ignored(?)
- `onClick` updates a state variable, `triggerNext`
- `triggerNext` is passed down into `NodeNextButton` as a prop
- The `useEffect` hook calls `onNext` as a side-effect of `triggerNext` changing, as long as `triggerNext` doesn't equal 0. This function call is what actually changes to the next page.
- `onNext`  is provided by a HOC inside `NodeModal` so it also is passed down into `NodeNextButton` as a prop.

This _seems_ like a _lot_ of steps and events propagating back and forth between these two components just to handle the `next` button click. I'm having trouble seeing why the `triggerNext` variable is needed. 

Just to see what would happen, I removed this `useEffect` hook and just passed `onNext` into the `onClick` property of the button directly and as far as I can tell everything still works? :confused:

Reply:
@jakemcdermott an astute observation.  I worked on this component with @keithjgrant before the break I believe.  This component is in place to solve a single problem: Adding steps to the workflow wizard dynamically.  In a previous iteration of this PR, the wizard steps were a bit different:

Step 1: Select the edge type
Step 2: Select the node type
Step 3: (Depended on what you selected from step 3 and is conditionally included in the steps array based on a state variable.)

The problem that we were running in to was that when the user clicked next, the wizard's `onNext` function was getting called _before_ the state variable was through updating (and steps array subsequently updated).  So the user would not get navigated to the dynamically added 3rd step but rather, the wizard would close.

This isn't immediately apparent the way the PR currently sits because the UX has changed slightly and we no longer dynamically add that 3rd step but it has been combined with step 2.  I didn't remove this component however because I think we will need this behavior as soon as we add prompting in.  The prompting steps will need to be dynamically added and I think we'll run into the same problem that this component solves.